components/brave_wallet/common/eth_abi_utils.cc

```diff
@@ -203,6 +208,67 @@ absl::optional<std::vector<std::string>> ExtractStringArray(Span string_array) {
   return result;
 }
 
+absl::optional<std::pair<bool, std::vector<uint8_t>>> ExtractBoolAndBytes(
+    Span data) {
+  auto bool_row = ExtractRow(data, 0);
+  if (!bool_row) {
+    return absl::nullopt;
+  }
+
+  auto bytes = ExtractBytesFromTuple(data, 1);
+  if (!bytes) {
+    return absl::nullopt;
+  }
+
+  return std::make_pair(BytesToBool(*bool_row), std::move(*bytes));
+}
+absl::optional<std::vector<std::pair<bool, std::vector<uint8_t>>>>
+ExtractBoolBytesTupleArray(Span data) {
+  // Get the length of the array.
+  auto array_size_offset_row = ExtractRow(data, 0);
+  if (!array_size_offset_row) {
+    return absl::nullopt;
+  }
+  absl::optional<size_t> offset = BytesToSize(*array_size_offset_row);
+  if (!offset || *offset > data.size()) {
+    return absl::nullopt;
+  }
+  auto array_size_row = ExtractRow(data, (*offset) / kRowLength);
+  if (!array_size_row) {
+    return absl::nullopt;
+  }
+  auto array_size = BytesToSize(*array_size_row);
+  if (!array_size || *array_size > PaddedRowCount(data.size())) {
+    return absl::nullopt;
+  }
+
+  // Increment the offset to the next row since size has been read.
+  *offset += kRowLength;
+
+  // Loop through each tuple in the array and decode.
+  std::vector<std::pair<bool, std::vector<uint8_t>>> result;
+  result.reserve(*array_size);
+  for (auto i = 0u; i < *array_size; ++i) {
+    auto tuple_offset_row =
+        ExtractRow(data, (*offset + (kRowLength * i)) / kRowLength);
```

Comment:
offset already points to start of array of heads, so
`ExtractRow(*offset, i)`

Reply:
Actually everything in this loop should be done in context of `offset`(maybe needs better name actually), but not 'data'. Parsing is designed with idea of being recursive and every part of structure should be parsed from ist own chunk.

Reply:
I misread above what offset is. Actually that should refer(with a better name) to a subspan of data from which array size trimmed. See ExtractStringArray for example